src/lightning/data/cache/dataloader.py

```diff
@@ -0,0 +1,196 @@
+# Copyright The Lightning AI team.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+from datetime import datetime
+from typing import Any, Optional
+
+import torch
+from torch.utils.data import Dataset, IterableDataset
+from torch.utils.data._utils.collate import default_collate
+from torch.utils.data.dataloader import (
+    DataLoader,
+    _BaseDataLoaderIter,
+    _MultiProcessingDataLoaderIter,
+    _SingleProcessDataLoaderIter,
+)
+
+from lightning.data.cache import Cache
+from lightning.data.cache.pytree import tree_flatten
+from lightning.data.cache.sampler import CacheBatchSampler
+from lightning.data.datasets.env import _DistributedEnv
+
+logger = logging.Logger(__name__)
+
+
+def _equal_items(data_1: Any, data_2: Any) -> bool:
+    data_1_flattened, _ = tree_flatten(data_1)
+    data_2_flattened, _ = tree_flatten(data_2)
+
+    if len(data_1_flattened) != len(data_2_flattened):
+        return False
+
+    return all(_equal_item(d1, d2) for d1, d2 in zip(data_1_flattened, data_2_flattened))
+
+
+def _equal_item(d1, d2) -> bool:
+    if not isinstance(d1, type(d2)):
+        raise False
+    if isinstance(d1, torch.Tensor) and isinstance(d2, torch.Tensor):
+        return torch.equal(d1, d2)
+    return d1 == d2
+
+
+class CacheDataset(Dataset):
+    def __init__(
+        self, dataset: Dataset, cache_dir: Optional[str], chunk_size: Optional[int], compression: Optional[str]
+    ):
+        self._datataset = dataset
+        if cache_dir is None:
+            cache_dir = os.path.join(os.getcwd(), "cache_dir", datetime.now().strftime("%m-%d-%Y-%H-%M"))
+            os.makedirs(cache_dir, exist_ok=True)
```

Comment:
I think the cache dir should be required. This default here will result in race conditions in a distributed setting at the minute boundary.

Reply:
I agree with @awaelchli. Either increase the resolution (probably even up to milliseconds in case of distributed) or (better) make it required

Reply:
I think it depends. Do you want to share the cache between processes or not? If sharing is desired, then the resolution should not be increased and instead a deterministic name chosen. Otherwise, you end up with different directories. If sharing is not desired, then the folders should probably use the rank to differentiate the folder name.

Reply:
Sure. Let's make it require.